pyvrp/search/neighbourhood.py

```diff
@@ -82,13 +82,21 @@ def compute_neighbours(
     if params.symmetric_proximity:
         proximity = np.minimum(proximity, proximity.T)
 
-    n = len(proximity)
-    k = min(params.nb_granular, data.num_clients - 1)  # excl. self
+    for group in data.groups():
+        if group.mutually_exclusive:
+            # Clients in mutually exclusive groups cannot neighbour each other,
+            # since only one of them can be in the solution at any given time.
+            # We use max float, not infty, to ensure these clients are ordered
+            # before the depots: we want to avoid same group neighbours, but it
+            # is not problematic if we need to have them.
+            idcs = np.ix_(group.clients, group.clients)
+            proximity[idcs] = np.finfo(np.float64).max
```

Comment:
> In other words, should we exclude neighbors that have a proximity of `np.inf`? I think we have to do this, because when both the depot and clients of the same group have a proximity of `np.inf`, then what can happen is that the depot becomes a neighbor.

`np.inf` is not the same as `np.info(np.float64).max`. In particular, `np.info(np.float64).max < np.inf`, and a depot can thus never become a neighbour.